### PR TITLE
Filtering empty names with naive approach

### PR DESCRIPTION
**Pull request by CaptainTrunky**
Connects to ros2/ros2#489

A pull request for #489 basic implementation that allocates a temporary buffer for node names and filters out all empty names. It's not properly tested yet and additional unit tests are required. So next steps are:

- [ ] add tests (ros2/system_tests#374)

- [ ] add optimal implementation that does not require additional buffer and copying data

PS
had to reopen PR because rebasing all the 6+ month old changes  is way too painful to be productive.

Signed-off-by: captaintrunky <captaintrunky@gmail.com>

### CODE REVIEW

**Comment by CaptainTrunky** [comment]
Finally I've finished merging with current master.

File: rmw_connext_shared_cpp/src/node_names.cpp [hunk 1]
```diff
@@ -96,27 +106,28 @@ get_node_names(
 
       if (name_found != map.end()) {
         name = std::string(name_found->second.begin(), name_found->second.end());
-      }
-      if (name_found != map.end()) {
```
Comment (jacobperron) [inline rmw_connext_shared_cpp/src/node_names.cpp]: Looks like this was previously a bug. Instead of removing the condition, I think it should be:

```c++
if (ns_found != map.end()) {
```

File: rmw_connext_shared_cpp/src/node_names.cpp [hunk 2]
```diff
@@ -96,27 +106,28 @@ get_node_names(
...
         namespace_ = std::string(ns_found->second.begin(), ns_found->second.end());
       }
+
+      // ignore discovered participants without a name
```
Comment (jacobperron) [inline rmw_connext_shared_cpp/src/node_names.cpp]: I think this comment isn't necessary. The block below is not ignoring discovered participants, it's using the participant name instead of the name from user data in the case that is is empty.

File: rmw_connext_shared_cpp/src/node_names.cpp [hunk 3]
```diff
@@ -96,27 +106,28 @@ get_node_names(
...
-      node_names->data[i] = nullptr;
+      tmp_names_list.data[i] = nullptr;
       node_namespaces->data[i] = nullptr;
       continue;
     }
 
-    node_names->data[i] = rcutils_strdup(name.c_str(), allocator);
-    if (!node_names->data[i]) {
-      RMW_SET_ERROR_MSG("could not allocate memory for node name");
+    tmp_names_list.data[named_nodes_num] = rcutils_strdup(name.c_str(), allocator);
```
Comment (jacobperron) [inline rmw_connext_shared_cpp/src/node_names.cpp]: I think this should use `i`, as we're using it on the following line.
```suggestion
    tmp_names_list.data[i] = rcutils_strdup(name.c_str(), allocator);
```

File: rmw_connext_shared_cpp/src/node_names.cpp [hunk 4]
```diff
@@ -96,27 +106,28 @@ get_node_names(
...
+    if (!tmp_names_list.data[i]) {
+      RMW_SET_ERROR_MSG("could not allocate memory for node namespace");
```
Comment (jacobperron) [inline rmw_connext_shared_cpp/src/node_names.cpp]:
```suggestion
      RMW_SET_ERROR_MSG("could not allocate memory for node name");
```

File: rmw_connext_shared_cpp/src/node_names.cpp [hunk 5]
```diff
@@ -125,26 +136,63 @@ get_node_names(
...
+  // TODO: implement functionality that allocates just enough memory for the output buffer,
+  // so using additional buffer would be unnessecary 
+  rcutils_ret = rcutils_string_array_init(node_names, named_nodes_num, &allocator);
+  if (rcutils_ret != RCUTILS_RET_OK) {
+    RMW_SET_ERROR_MSG(rcutils_get_error_string().str);
+    return rmw_convert_rcutils_ret_to_rmw_ret(rcutils_ret);
+  }
+
+  for (auto i = 0; i < named_nodes_num; ++i) {
+    node_names->data[i] = rcutils_strdup(tmp_names_list.data[i], allocator);
```
Comment (jacobperron) [inline rmw_connext_shared_cpp/src/node_names.cpp]: This logic looks incorrect. It's possible there is an empty name in the middle of `tmp_names_list`, in which case we're copying `nullptr` (and missing a potential valid name at the end of the list).

The for-loop should loop up to `length` and then inside we should guard against `nullptr`:

```c++
size_t j = 0;
for (auto i = 0; i < length; ++i) {
  if (tmp_names_list.data[i]) {
    node_names->data[j] = rcutils_strdup(tmp_names_list.data[i], allocator);
  }
  ++j;
}
```

File: rmw_connext_shared_cpp/src/node_names.cpp [hunk 6]
```diff
@@ -125,26 +136,63 @@ get_node_names(
...
+    tmp_names_list.data[i] = nullptr;
   }
 
+  // prepare actual output without empty names
+  // TODO: implement functionality that allocates just enough memory for the output buffer,
+  // so using additional buffer would be unnessecary 
+  rcutils_ret = rcutils_string_array_init(node_names, named_nodes_num, &allocator);
```
Comment (jacobperron) [inline rmw_connext_shared_cpp/src/node_names.cpp]: This block and the next are duplicated.

File: rmw_connext_shared_cpp/src/node_names.cpp [hunk 7]
```diff
@@ -125,26 +136,63 @@ get_node_names(
...
+
 fail:
   if (node_names) {
     rcutils_ret = rcutils_string_array_fini(node_names);
     if (rcutils_ret != RCUTILS_RET_OK) {
-      RCUTILS_LOG_ERROR_NAMED(
-        "rmw_connext_cpp",
-        "failed to cleanup during error handling: %s", rcutils_get_error_string().str);
-      rcutils_reset_error();
+      RMW_SET_ERROR_MSG(rcutils_get_error_string().str);
```
Comment (jacobperron) [inline rmw_connext_shared_cpp/src/node_names.cpp]: In all cases where we `goto fail`, the error message is already set. You can undo this change.

File: rmw_connext_shared_cpp/src/node_names.cpp [hunk 8]
```diff
@@ -125,26 +136,63 @@ get_node_names(
...
+      return rmw_convert_rcutils_ret_to_rmw_ret(rcutils_ret);
     }
   }
+  rcutils_ret = rcutils_string_array_fini(&tmp_names_list);
+  if (rcutils_ret != RCUTILS_RET_OK) {
+    RMW_SET_ERROR_MSG(rcutils_get_error_string().str);
+    return rmw_convert_rcutils_ret_to_rmw_ret(rcutils_ret);
```
Comment (jacobperron) [inline rmw_connext_shared_cpp/src/node_names.cpp]: Instead of setting an error message and returning, log an error and reset the error:

```c++
RCUTILS_LOG_ERROR_NAMED(
  "rmw_connext_cpp",
  "failed to cleanup during error handling: %s", rcutils_get_error_string().str);
rcutils_reset_error();
```

File: rmw_connext_shared_cpp/src/node_names.cpp [hunk 9]
```diff
@@ -125,26 +136,63 @@ get_node_names(
...
+  }
   if (node_namespaces) {
     rcutils_ret = rcutils_string_array_fini(node_namespaces);
     if (rcutils_ret != RCUTILS_RET_OK) {
-      RCUTILS_LOG_ERROR_NAMED(
-        "rmw_connext_cpp",
-        "failed to cleanup during error handling: %s", rcutils_get_error_string().str);
-      rcutils_reset_error();
+      RMW_SET_ERROR_MSG(rcutils_get_error_string().str);
+      return rmw_convert_rcutils_ret_to_rmw_ret(rcutils_ret);
```
Comment (jacobperron) [inline rmw_connext_shared_cpp/src/node_names.cpp]: Undo this change.

File: rmw_connext_shared_cpp/src/node_names.cpp [hunk 10]
```diff
@@ -45,41 +45,51 @@ get_node_names(
...
+  rcutils_ret_t rcutils_ret = rcutils_string_array_init(node_namespaces, length, &allocator);
+
   if (rcutils_ret != RCUTILS_RET_OK) {
     RMW_SET_ERROR_MSG(rcutils_get_error_string().str);
     return rmw_convert_rcutils_ret_to_rmw_ret(rcutils_ret);
   }
-  rcutils_ret = rcutils_string_array_init(node_namespaces, length, &allocator);
+
+  // Pre-allocate temporary buffer for all node names
+  // Nodes that are not created by ROS2 could provide empty names
```
Comment (jacobperron) [inline rmw_connext_shared_cpp/src/node_names.cpp]: nitpick: "ROS 2" (or just "ROS").

**Comment by nuclearsandwich** [comment]
@CaptainTrunky do you have plans to resume iterating on this PR? Thanks.

**Comment by jacobperron** [comment]
@CaptainTrunky Friendly ping :slightly_smiling_face: Let us know if you'd like to keep working on this.

**Comment by CaptainTrunky** [comment]
Sorry, I've been busy with my job. I'll try to resolve current issues within next two weeks.

File: rmw_connext_shared_cpp/src/node_names.cpp (hunk 3 above)
Comment (CaptainTrunky) [reply rmw_connext_shared_cpp/src/node_names.cpp]: I guess that that's not right. My idea was to overallocate `tmp_names_list`, so it might be identical to `node_names`. Next, I add node names to tmp_names_list only if a name is not empty. Thus, `named_nodes_num` tracks number of non-empty names so far. After traversing the whole list of names, `tmp_names_list` contains valid names and (possibly) nullptrs at the end. We can safely truncate them later. Is this logic applicable?

Comment (jacobperron) [reply rmw_connext_shared_cpp/src/node_names.cpp]: Okay, makes sense. We should do the same logic for the namespaces as well, no?

File: rmw_connext_shared_cpp/src/node_names.cpp (hunk 5 above)
Comment (jacobperron) [reply rmw_connext_shared_cpp/src/node_names.cpp]: I overlooked the logic above (https://github.com/ros2/rmw_connext/pull/362/files#r327540360), this loop looks okay.

File: rmw_connext_shared_cpp/src/node_names.cpp (hunk 3 above)
Comment (CaptainTrunky) [reply rmw_connext_shared_cpp/src/node_names.cpp]: I think so. This function treats names and namespaces in almost identical way.

**Comment by dirk-thomas** [comment]
@CaptainTrunky Another ping

**Comment by CaptainTrunky** [comment]
I've added support for namespaces. Plan to to test it on Wednesday and push final version.

File: rmw_connext_shared_cpp/src/node_names.cpp (hunk 9 above)
Comment (CaptainTrunky) [reply rmw_connext_shared_cpp/src/node_names.cpp]: Done

File: rmw_connext_shared_cpp/src/node_names.cpp (hunk 10 above)
Comment (CaptainTrunky) [reply rmw_connext_shared_cpp/src/node_names.cpp]: Done

File: rmw_connext_shared_cpp/src/node_names.cpp (hunk 8 above)
Comment (CaptainTrunky) [reply rmw_connext_shared_cpp/src/node_names.cpp]: Done

File: rmw_connext_shared_cpp/src/node_names.cpp (hunk 7 above)
Comment (CaptainTrunky) [reply rmw_connext_shared_cpp/src/node_names.cpp]: Done

File: rmw_connext_shared_cpp/src/node_names.cpp (hunk 6 above)
Comment (CaptainTrunky) [reply rmw_connext_shared_cpp/src/node_names.cpp]: Done

File: rmw_connext_shared_cpp/src/node_names.cpp (hunk 4 above)
Comment (CaptainTrunky) [reply rmw_connext_shared_cpp/src/node_names.cpp]: Done

File: rmw_connext_shared_cpp/src/node_names.cpp (hunk 3 above)
Comment (CaptainTrunky) [reply rmw_connext_shared_cpp/src/node_names.cpp]: Done

File: rmw_connext_shared_cpp/src/node_names.cpp (hunk 2 above)
Comment (CaptainTrunky) [reply rmw_connext_shared_cpp/src/node_names.cpp]: Done

File: rmw_connext_shared_cpp/src/node_names.cpp (hunk 1 above)
Comment (CaptainTrunky) [reply rmw_connext_shared_cpp/src/node_names.cpp]: Done

File: rmw_connext_shared_cpp/src/node_names.cpp [hunk 11]
```diff
@@ -90,44 +104,100 @@ get_node_names(
...
+    ++named_namespaces_num;
   }
 
+  // prepare actual output without empty names
+  // TODO: implement functionality that allocates just enough memory for the output buffer,
+  // so using additional buffer would be unnessecary 
+  rcutils_ret = rcutils_string_array_init(node_names, named_nodes_num, &allocator);
+  if (rcutils_ret != RCUTILS_RET_OK) {
+    RMW_SET_ERROR_MSG(rcutils_get_error_string().str);
+    return rmw_convert_rcutils_ret_to_rmw_ret(rcutils_ret);
```
Comment (jacobperron) [inline rmw_connext_shared_cpp/src/node_names.cpp]: We should `goto fail;` here so we can cleanup.

File: rmw_connext_shared_cpp/src/node_names.cpp [hunk 12]
```diff
@@ -90,44 +104,100 @@ get_node_names(
...
-      node_names->data[i] = nullptr;
-      node_namespaces->data[i] = nullptr;
+      tmp_names_list.data[i] = nullptr;
+      tmp_namespaces_list.data[i] = nullptr;
       continue;
     }
 
-    node_names->data[i] = rcutils_strdup(name.c_str(), allocator);
-    if (!node_names->data[i]) {
+    tmp_names_list.data[named_nodes_num] = rcutils_strdup(name.c_str(), allocator);
```
Comment (jacobperron) [inline rmw_connext_shared_cpp/src/node_names.cpp]: This will overwrite the first element of `tmp_names_list` that was set at L93. I think `named_nodes_num` should be initialized to `1`. Same for `named_namespaces_num`.

File: rmw_connext_shared_cpp/src/node_names.cpp [hunk 13]
```diff
@@ -90,44 +104,100 @@ get_node_names(
...
+    RMW_SET_ERROR_MSG(rcutils_get_error_string().str);
+    return rmw_convert_rcutils_ret_to_rmw_ret(rcutils_ret);
+  }
+
+  for (auto i = 0; i < named_nodes_num; ++i) {
+    node_names->data[i] = rcutils_strdup(tmp_names_list.data[i], allocator);
+    tmp_names_list.data[i] = nullptr;
+  }
+
+  for (auto i = 0; i < named_nodes_num; ++i) {
```
Comment (jacobperron) [inline rmw_connext_shared_cpp/src/node_names.cpp]: I guess this loop is leftover from debugging and can be removed.

File: rmw_connext_shared_cpp/src/node_names.cpp [hunk 14]
```diff
@@ -12,6 +12,7 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+#include <iostream>
```
Comment (jacobperron) [inline rmw_connext_shared_cpp/src/node_names.cpp]: I think this can be removed.

File: rmw_connext_shared_cpp/src/node_names.cpp [hunk 15]
```diff
@@ -90,44 +104,100 @@ get_node_names(
...
       RMW_SET_ERROR_MSG("could not allocate memory for node name");
       goto fail;
     }
 
-    node_namespaces->data[i] = rcutils_strdup(namespace_.c_str(), allocator);
-    if (!node_namespaces->data[i]) {
+    ++named_nodes_num;
+    
+    if (namespace_.empty()) {
+      tmp_namespaces_list.data[i] = nullptr;
```
Comment (jacobperron) [inline rmw_connext_shared_cpp/src/node_names.cpp]: This is concerning, since we expect that the output names and namespaces lists will be the same size representing a one-to-one mapping for nodes ([docs](https://github.com/ros2/rmw/blob/813b94ddd5650444b312304ad156f3b5d9f05e39/rmw/include/rmw/rmw.h#L1015-L1018)). It is possible that if we have empty namespaces, but not empty names, that we will end up with lists of different sizes and lose the mapping. In other words, we want to ensure that `named_nodes_num == named_namespaces_num`.

I think that if the namespace is empty, we can let it be empty and not filter it like we do with the node names. We can probably just remove this block.

File: rmw_connext_shared_cpp/src/node_names.cpp [hunk 16]
```diff
@@ -90,44 +104,100 @@ get_node_names(
...
+    tmp_names_list.data[i] = nullptr;
+  }
+
+  for (auto i = 0; i < named_nodes_num; ++i) {
+    std::cerr << tmp_names_list.data[i] << '\n';
+  }
+  rcutils_ret = rcutils_string_array_fini(&tmp_names_list);
+  if (rcutils_ret != RCUTILS_RET_OK) {
+    RMW_SET_ERROR_MSG(rcutils_get_error_string().str);
+    return rmw_convert_rcutils_ret_to_rmw_ret(rcutils_ret);
```
Comment (jacobperron) [inline rmw_connext_shared_cpp/src/node_names.cpp]: Be careful returning too early, we are leaking memory since `tmp_namespaces_list` isn't being finalized.
I suggest refactoring so that we only return once at the end of the function. We can rename the `fail` label to `cleanup`, and whenever there is an error, we can set an error message, set a return code, and `goto cleanup`.

E.g. 

```c
rmw_ret_t final_ret = RMW_RET_OK;  // put this near the top after the initial error handling (e.g. L65)

// ...

if (rcutils_ret_ != RCUTILS_RET_OK) {
  RMW_SET_ERROR_MSG(rcutils_get_error_string().str);
  final_ret = rmw_convert_rcutils_ret_to_rmw_ret(rcutils_ret);
  goto cleanup;
}

// ...

cleanup:
// ... cleanup logic
return final_ret;
```

File: rmw_connext_shared_cpp/src/node_names.cpp (hunk 15 above)
Comment (jacobperron) [reply rmw_connext_shared_cpp/src/node_names.cpp]: In fact, we might remove the variable `named_namespaces_num` altogether since it should always be the same as `named_nodes_num`.

File: rmw_connext_shared_cpp/src/node_names.cpp (hunk 16 above)
Comment (CaptainTrunky) [reply rmw_connext_shared_cpp/src/node_names.cpp]: Done

File: rmw_connext_shared_cpp/src/node_names.cpp (hunk 15 above)
Comment (CaptainTrunky) [reply rmw_connext_shared_cpp/src/node_names.cpp]: Done

File: rmw_connext_shared_cpp/src/node_names.cpp (hunk 14 above)
Comment (CaptainTrunky) [reply rmw_connext_shared_cpp/src/node_names.cpp]: Done

File: rmw_connext_shared_cpp/src/node_names.cpp (hunk 13 above)
Comment (CaptainTrunky) [reply rmw_connext_shared_cpp/src/node_names.cpp]: Done

File: rmw_connext_shared_cpp/src/node_names.cpp (hunk 12 above)
Comment (CaptainTrunky) [reply rmw_connext_shared_cpp/src/node_names.cpp]: Done

File: rmw_connext_shared_cpp/src/node_names.cpp (hunk 11 above)
Comment (CaptainTrunky) [reply rmw_connext_shared_cpp/src/node_names.cpp]: Done

**Comment by jacobperron** [comment]
Sorry for the delay in getting back to this. I've rebased on master and added a small fix (70e6973) and resolved linter errors (68f59bb).

Running CI for all platforms to see where we stand:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/8737/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/4566/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/7139/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/8651/)

File: rmw_connext_shared_cpp/src/node_names.cpp [hunk 17]
```diff
@@ -48,38 +48,56 @@ get_node_names(
...
-  if (!node_names->data[0]) {
-    RMW_SET_ERROR_MSG("could not allocate memory for node name");
-    return RMW_RET_BAD_ALLOC;
-  }
-  node_namespaces->data[0] = rcutils_strdup(node->namespace_, allocator);
-  if (!node_names->data[0]) {
-    RMW_SET_ERROR_MSG("could not allocate memory for node namespace");
-    return RMW_RET_BAD_ALLOC;
+  // add yourself
+  tmp_names_list.data[0] = rcutils_strdup(participant_qos.participant_name.name, allocator);
```
Comment (dirk-thomas) [inline rmw_connext_shared_cpp/src/node_names.cpp]: This only populates the `name` bub not the `namespace` of the current node. As a result a test in `rcl` is failing since this PR was merged.

The CI builds ran for this change also didn't cover `rcl` as far as I can see. Please address this regression as soon as possible.

Comment (jacobperron) [reply rmw_connext_shared_cpp/src/node_names.cpp]: My mistake; I only ran test for this package and test_communication. I'll fix this.

Comment (jacobperron) [reply rmw_connext_shared_cpp/src/node_names.cpp]: See https://github.com/ros2/rmw_connext/pull/384 for a fix.

File: rmw_connext_shared_cpp/src/node_names.cpp [hunk 18]
```diff
@@ -48,38 +48,56 @@ get_node_names(
...
   rcutils_allocator_t allocator = rcutils_get_default_allocator();
-  rcutils_ret_t rcutils_ret = rcutils_string_array_init(node_names, length, &allocator);
+
+  // Pre-allocate temporary buffer for all node names & namespaces
+  // Nodes that are not created by ROS 2 could provide empty names
+  // Such names should not be returned
+  rcutils_string_array_t tmp_names_list = rcutils_get_zero_initialized_string_array();
+  rcutils_string_array_t tmp_namespaces_list = rcutils_get_zero_initialized_string_array();
+
+  int named_nodes_num = 1;
```
Comment (ivanpauno) [inline rmw_connext_shared_cpp/src/node_names.cpp]: why this starts from 1?

File: rmw_connext_shared_cpp/src/node_names.cpp [hunk 19]
```diff
@@ -90,45 +108,93 @@ get_node_names(
...
+
+  if (rcutils_ret != RCUTILS_RET_OK) {
+    RMW_SET_ERROR_MSG("could not allocate memory for node_namespaces output");
+    final_ret = rmw_convert_rcutils_ret_to_rmw_ret(rcutils_ret);
+    goto cleanup;
+  }
+
+  for (auto i = 0; i < named_nodes_num; ++i) {
+    node_names->data[i] = rcutils_strdup(tmp_names_list.data[i], allocator);
+    tmp_names_list.data[i] = nullptr;
```
Comment (ivanpauno) [inline rmw_connext_shared_cpp/src/node_names.cpp]: isn't this leaking memory?

File: rmw_connext_shared_cpp/src/node_names.cpp (hunk 18 above)
Comment (jacobperron) [reply rmw_connext_shared_cpp/src/node_names.cpp]: The first element is set to the current node name at L96. I guess for clarity this could have started at zero and then been incremented after L96.

File: rmw_connext_shared_cpp/src/node_names.cpp (hunk 19 above)
Comment (jacobperron) [reply rmw_connext_shared_cpp/src/node_names.cpp]: Yes, I suppose it does. Good catch! I don't think this line or the other at L178 are necessary.

See https://github.com/ros2/rmw_connext/pull/395 for a fix.

File: rmw_connext_shared_cpp/src/node_names.cpp (hunk 18 above)
Comment (ivanpauno) [reply rmw_connext_shared_cpp/src/node_names.cpp]: Yes, didn't realize.
The logic is a bit confusing.
